packages/langium/test/grammar/grammar-validator.test.ts

```diff
@@ -1031,6 +1031,104 @@ describe('Assignments with = instead of +=', () => {
         expect(validation.diagnostics[0].message).toBe(getMessage('left'));
         expect(validation.diagnostics[1].message).toBe(getMessage('left'));
     });
+
+    test('rewrite actions inside loop #1756 (complete examples, slightly adapted)', async () => {
+        const validation = await validate(getGrammar(`
+            entry Model:
+                expr=Expression;
+
+            Expression: Equality;
+
+            Equality infers Expression:
+                Literal ( ( {infer Equals.left=current} '==' | {infer NotEquals.left=current} '!=' )  right=Literal)*;
+
+            Literal: value=ID;
+        `));
+        expectNoIssues(validation);
+    });
+
+    test('rewrite actions inside loop #1756 (with a single alternative only)', async () => {
+        const validation = await validate(getGrammar(`
+            entry Model:
+                expr=Expression;
+
+            Expression: Equality;
+
+            Equality infers Expression:
+                Literal ( {infer Equals.left=current} '==' right=Literal)*;
+
+            Literal: value=ID;
+        `));
+        expectNoIssues(validation);
+    });
+
+    test('rewrite actions inside loop #1756 (single alternative in non-empty group)', async () => {
+        const validation = await validate(getGrammar(`
+            entry Model:
+                expr=Expression;
+
+            Expression: Equality;
+
+            Equality infers Expression:
+                Literal ( {infer Equals.left=current} 'nonemptygroup' '==' right=Literal)*;
+
+            Literal: value=ID;
+        `));
+        expectNoIssues(validation);
+    });
+
+    test('rewrite actions inside loop #1756 (single alternative in non-empty grouped group)', async () => {
+        const validation = await validate(getGrammar(`
+            entry Model:
+                expr=Expression;
+
+            Expression: Equality;
+
+            Equality infers Expression:
+                Literal ( ( {infer Equals.left=current} 'nonemptygroup' ) '==' right=Literal)*;
+
+            Literal: value=ID;
+        `));
+        expectNoIssues(validation);
+    });
+
+    test('rewrite actions inside loop #1756 (only a single alternative creates a new object)', async () => {
+        // Since we assume the worst case and since 'right' is assigned twice, if the alternative without the rewrite action is used, we expect a warning here as well!
+        const validation = await validate(getGrammar(`
+            entry Model:
+                expr=Expression;
+
+            Expression: Equality;
+
+            Equality infers Expression:
+                Literal ( ( {infer Equals.left=current} '==' | '!=' )  right=Literal)*;
```

Comment:
Very nice case, fully agree to bring up the warning here, as that indicates that the grammar design is probably sub optimal.